File: android/src/main/java/expo/modules/adaptiveglass/renderers/AcrylicRenderer.kt
package expo.modules.adaptiveglass.renderers

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.RadialGradient
import android.graphics.RectF
import android.graphics.Shader

// Tint, sheen, border and highlight over the blur (or alone). Gradients only rebuild on size or
// color changes.
class AcrylicRenderer(private val density: Float) {
  data class Style(
    val dark: Boolean,
    val tint: Int?,
    val intensity: Float,
    val opaque: Boolean,
    val minimal: Boolean,
    // 0 frosted to 1 clear
    val clarity: Float = 0f,
    val ultra: Boolean = false,
  )

  private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val sheenPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }
  // stacked strokes, brightest at the rim and fading inward. The lit edge of thick glass at ultra
  private val bezelPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }
  private val bezelRect = RectF()
  private val specPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val rect = RectF()
  private val specMatrix = Matrix()

  private var style: Style? = null
  private var shaderW = -1
  private var shaderH = -1
  private var liveFillAlpha = 0f
  private var acrylicFillAlpha = 0f
  private var baseColor = Color.WHITE
  private var light = Color.WHITE

  fun configure(next: Style) {
    if (next == style) return
    style = next
    val i = next.intensity
    baseColor = next.tint?.let { it or (0xFF shl 24) }
      ?: if (next.dark) Color.rgb(30, 30, 34) else Color.rgb(247, 247, 250)
    // neutral frost clears with clarity. A colour tint gets stronger instead, or the frost and a
    // sharper background wash it out
    val c = if (next.tint != null) 0f else next.clarity
    liveFillAlpha = if (next.tint != null) {
      (0.25f + 0.35f * i) * (1f + 0.4f * next.clarity)
    } else {
      ((if (next.dark) 0.14f else 0.1f) + 0.22f * i) * (1f - 0.6f * c)
    }
    // no blur under acrylic, so it only clears so far before text behind gets hard to read past
    acrylicFillAlpha = if (next.opaque) 0.97f else 0.62f + 0.28f * i - 0.3f * c
    light = highlight(next.tint, next.dark)
    shaderW = -1 // force gradient rebuild
  }

  // liveMix: 1 with live blur underneath, 0 for pure acrylic
  fun draw(canvas: Canvas, w: Int, h: Int, radius: Float, liveMix: Float, specAlpha: Float, sx: Float, sy: Float) {
    val s = style ?: return
    if (w != shaderW || h != shaderH) rebuildShaders(w, h, s)
    rect.set(0f, 0f, w.toFloat(), h.toFloat())

    val fillAlpha = acrylicFillAlpha + (liveFillAlpha - acrylicFillAlpha) * liveMix
    fillPaint.color = withAlpha(baseColor, fillAlpha)
    canvas.drawRoundRect(rect, radius, radius, fillPaint)

    if (!s.minimal && !s.opaque) canvas.drawRoundRect(rect, radius, radius, sheenPaint)

    if (specAlpha > 0.001f) {
      val r = maxOf(w, h) * 0.6f
      specMatrix.setScale(r, r)
      specMatrix.postTranslate(sx, sy)
      specPaint.shader?.setLocalMatrix(specMatrix)
      specPaint.alpha = (specAlpha * 255).toInt()
      canvas.drawRoundRect(rect, radius, radius, specPaint)
    }

    if (s.ultra && !s.opaque) {
      for (i in BEZEL_WIDTHS.indices) {
        val bw = BEZEL_WIDTHS[i] * density
        bezelPaint.strokeWidth = bw
        bezelPaint.color = withAlpha(light, BEZEL_ALPHAS[i] * (if (s.dark) 0.7f else 1f))
        bezelRect.set(0f, 0f, w.toFloat(), h.toFloat())
        bezelRect.inset(bw / 2, bw / 2)
        canvas.drawRoundRect(bezelRect, maxOf(0f, radius - bw / 2), maxOf(0f, radius - bw / 2), bezelPaint)
      }
    }

    val half = borderPaint.strokeWidth / 2
    rect.inset(half, half)
    canvas.drawRoundRect(rect, radius - half, radius - half, borderPaint)
  }

  private fun rebuildShaders(w: Int, h: Int, s: Style) {
    shaderW = w
    shaderH = h
    // highlights stay at any clarity, they're what still reads as glass when clear
    val sheenAlpha = (if (s.dark) 0.1f else 0.2f) * (0.5f + maxOf(s.intensity, 0.5f))
    sheenPaint.shader = LinearGradient(
      0f, 0f, 0f, h * 0.6f,
      withAlpha(light, sheenAlpha), withAlpha(light, 0f), Shader.TileMode.CLAMP,
    )
    borderPaint.strokeWidth = maxOf(1f, 0.75f * density)
    // lit top-left only, a full rim reads as a white border on clear glass
    borderPaint.shader = if (s.opaque) {
      LinearGradient(0f, 0f, w.toFloat(), h.toFloat(), withAlpha(light, 0.9f), withAlpha(light, 0.5f), Shader.TileMode.CLAMP)
    } else if (s.ultra) {
      // ultra gets a soft rim all the way round like iOS 26 clear glass, brightest top-left
      val top = if (s.dark) 0.45f else 0.35f
      val rest = if (s.dark) 0.14f else 0.15f
      LinearGradient(
        0f, 0f, w.toFloat(), h.toFloat(),
        intArrayOf(withAlpha(light, top), withAlpha(light, rest), withAlpha(light, rest)),
        floatArrayOf(0f, 0.5f, 1f), Shader.TileMode.CLAMP,
      )
    } else {
      val top = (if (s.dark) 0.45f else 0.38f) * (if (s.minimal) 0.6f else 1f)
      LinearGradient(
        0f, 0f, w.toFloat(), h.toFloat(),
        intArrayOf(withAlpha(light, top), withAlpha(light, 0f), withAlpha(light, 0f)),
        floatArrayOf(0f, 0.45f, 1f), Shader.TileMode.CLAMP,
      )
    }
    // unit radius at the origin, moved with a local matrix so touches don't allocate
    specPaint.shader = RadialGradient(
      0f, 0f, 1f,
      withAlpha(light, if (s.dark) 0.32f else 0.55f), withAlpha(light, 0f), Shader.TileMode.CLAMP,
    )
  }

  companion object {
    private val BEZEL_WIDTHS = floatArrayOf(1.5f, 3.5f, 7f)
    // lower than iOS: the stacked strokes read stronger on Android at the same alpha
    private val BEZEL_ALPHAS = floatArrayOf(0.055f, 0.035f, 0.02f)
    fun withAlpha(color: Int, alpha: Float) =
      (color and 0x00FFFFFF) or ((alpha.coerceIn(0f, 1f) * 255).toInt() shl 24)

    // highlights lean toward the tint, not pure white
    private fun highlight(tint: Int?, dark: Boolean): Int {
      tint ?: return Color.WHITE
      val k = if (dark) 0.6f else 0.75f
      fun mix(c: Int) = (c + (255 - c) * k).toInt()
      return Color.rgb(mix(Color.red(tint)), mix(Color.green(tint)), mix(Color.blue(tint)))
    }
  }
}

File: ios/renderers/AcrylicRenderer.swift
import UIKit

/// Tint, sheen, hairline border and specular highlight, drawn over the blur or on their own.
/// Layers are built once, the border path only on size or radius changes.
final class AcrylicRenderer {
  enum Mode {
    case system, liveBlur, acrylic, opaque
  }

  struct Style: Equatable {
    var mode: Mode
    var dark: Bool
    var tint: UIColor?
    var intensity: CGFloat
    var minimal: Bool
    // 0 frosted to 1 clear
    var clarity: CGFloat = 0
    var ultra = false
    // set by the app, in place of the rim the glass lights itself
    var edge: UIColor?
    var edgeWidth: CGFloat = 0
  }

  let root = CALayer()
  private let fill = CALayer()
  private let sheen = CAGradientLayer()
  private let border = CAGradientLayer()
  // a bordered layer, not a stroked path: a UIBezierPath rounded rect curves differently from the
  // continuous clip and shows as a second edge inside the glass
  private let borderMask = CALayer()
  // stacked borders, brightest at the rim and fading inward. Fakes the lit edge of thick glass at
  // ultra, where iOS < 26 can't bend what's behind
  private let bezel = (0..<3).map { _ in CALayer() }
  private static let bezelWidths: [CGFloat] = [1.5, 3.5, 7]
  private static let bezelAlphas: [CGFloat] = [0.08, 0.05, 0.03]
  private let specular = CAGradientLayer()
  private var laidOutSize = CGSize.zero
  private var laidOutRadius: CGFloat = -1
  private var style: Style?

  init() {
    root.masksToBounds = true
    if #available(iOS 13.0, *) { root.cornerCurve = .continuous }
    sheen.startPoint = CGPoint(x: 0.5, y: 0)
    sheen.endPoint = CGPoint(x: 0.5, y: 0.6)
    // lit from the top-left, same as the Android shader
    border.startPoint = CGPoint(x: 0, y: 0)
    border.endPoint = CGPoint(x: 1, y: 1)
    borderMask.borderColor = UIColor.black.cgColor
    if #available(iOS 13.0, *) { borderMask.cornerCurve = .continuous }
    border.mask = borderMask
    specular.type = .radial
    specular.startPoint = CGPoint(x: 0.5, y: 0.5)
    specular.endPoint = CGPoint(x: 1, y: 1)
    specular.opacity = 0
    for layer in bezel {
      if #available(iOS 13.0, *) { layer.cornerCurve = .continuous }
      root.addSublayer(layer)
    }
    [fill, sheen, border, specular].forEach(root.addSublayer)
  }

  private var edgeWidth: CGFloat = 0

  func layout(bounds: CGRect, radius: CGFloat, scale: CGFloat) {
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    root.frame = bounds
    root.cornerRadius = radius
    fill.frame = bounds
    sheen.frame = bounds
    border.frame = bounds
    if bounds.size != laidOutSize || radius != laidOutRadius {
      laidOutSize = bounds.size
      laidOutRadius = radius
      let width = edgeWidth > 0 ? edgeWidth : max(1 / scale, 0.75)
      borderMask.frame = bounds
      borderMask.borderWidth = width
      borderMask.cornerRadius = radius
      for (i, layer) in bezel.enumerated() {
        layer.frame = bounds
        layer.cornerRadius = radius
        layer.borderWidth = Self.bezelWidths[i]
      }
      let d = max(bounds.width, bounds.height) * 0.9
      specular.bounds = CGRect(x: 0, y: 0, width: d, height: d)
    }
    CATransaction.commit()
  }

  func apply(_ next: Style, animated: Bool) {
    if next.edgeWidth != style?.edgeWidth {
      // the rim width is set during layout, force it to run again
      laidOutSize = .zero
    }
    guard next != style else { return }
    style = next
    CATransaction.begin()
    CATransaction.setDisableActions(!animated)
    CATransaction.setAnimationDuration(0.3)

    let i = next.intensity
    let neutral = next.dark ? UIColor(white: 0.12, alpha: 1) : UIColor(white: 0.97, alpha: 1)
    let base = next.tint ?? neutral
    let fillAlpha: CGFloat
    switch next.mode {
    case .system: fillAlpha = 0  // UIGlassEffect carries the tint itself
    // neutral frost clears with clarity. A colour tint gets stronger instead, or the frost and a
    // sharper background wash it out
    case .liveBlur:
      fillAlpha =
        next.tint == nil
        ? ((next.dark ? 0.12 : 0.08) + 0.22 * i) * (1 - 0.6 * next.clarity)
        : (0.25 + 0.35 * i) * (1 + 0.4 * next.clarity)
    // no blur under acrylic, so it only clears so far before text behind gets hard to read past
    case .acrylic: fillAlpha = 0.62 + 0.28 * i - (next.tint == nil ? 0.3 * next.clarity : 0)
    case .opaque: fillAlpha = 0.97
    }
    fill.backgroundColor = base.withAlphaComponent(fillAlpha).cgColor

    // highlights lean toward the tint instead of pure white
    let light = Self.highlight(tint: next.tint, dark: next.dark)
    let showDecor = next.mode != .system && !next.minimal
    // highlights stay at any clarity, they're what still reads as glass when clear
    let sheenAlpha: CGFloat = showDecor ? (next.dark ? 0.1 : 0.2) * (0.5 + max(i, 0.5)) : 0
    sheen.colors = [light.withAlphaComponent(sheenAlpha).cgColor, light.withAlphaComponent(0).cgColor]

    let bezelOn = next.ultra && next.mode != .system && next.mode != .opaque
    for (i, layer) in bezel.enumerated() {
      layer.borderColor = light.withAlphaComponent(bezelOn ? Self.bezelAlphas[i] * (next.dark ? 0.7 : 1) : 0).cgColor
    }
    if let edge = next.edge {
      border.colors = [edge.cgColor, edge.cgColor, edge.cgColor]
      border.locations = [0, 0.5, 1]
    } else if next.mode == .opaque {
      border.colors = [light.withAlphaComponent(0.9).cgColor, light.withAlphaComponent(0.5).cgColor]
      border.locations = nil
    } else if next.ultra {
      // ultra gets a soft rim all the way round like iOS 26 clear glass, brightest top-left
      let top: CGFloat = next.dark ? 0.45 : 0.35
      let rest: CGFloat = next.dark ? 0.14 : 0.15
      border.colors = [
        light.withAlphaComponent(top).cgColor, light.withAlphaComponent(rest).cgColor,
        light.withAlphaComponent(rest).cgColor,
      ]
      border.locations = [0, 0.5, 1]
    } else {
      // lit top-left only, a full rim reads as a white border on clear glass
      let top: CGFloat = next.dark ? 0.45 : 0.38
      border.colors = [
        light.withAlphaComponent(top).cgColor, light.withAlphaComponent(0).cgColor,
        light.withAlphaComponent(0).cgColor,
      ]
      border.locations = [0, 0.45, 1]
    }
    border.opacity = next.edge != nil ? 1 : (next.mode == .system ? 0 : (next.minimal ? 0.6 : 1))

    specular.colors = [
      light.withAlphaComponent(next.dark ? 0.32 : 0.55).cgColor,
      light.withAlphaComponent(0).cgColor,
    ]
    CATransaction.commit()
  }

  func setHighlight(at point: CGPoint?, visible: Bool) {
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    if let point { specular.position = point }
    CATransaction.commit()
    // opacity keeps its implicit fade
    specular.opacity = visible ? 1 : 0
  }

  private static func highlight(tint: UIColor?, dark: Bool) -> UIColor {
    guard let tint else { return .white }
    var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
    guard tint.getRed(&r, green: &g, blue: &b, alpha: &a) else { return .white }
    let k: CGFloat = dark ? 0.6 : 0.75
    return UIColor(red: r + (1 - r) * k, green: g + (1 - g) * k, blue: b + (1 - b) * k, alpha: 1)
  }
}
